Python/chapter01/1.3 - URLify/miguel_1.3_sol.py

```diff
@@ -0,0 +1,54 @@
+"""
+Python version 3.7.0
+1.3 - URLify
+Write a method to replace all spaces in a string with "%20".  You may assume that the string has sufficient
+space at the end to hold the additional characters, and that you are given the "true" length of the string.
+(Note: If implementing in Java, please use a character array so that you can perform this operation in place)
+"""
+import unittest
+from typing import Callable
+
+
+def urlify(s: str, true_length: int) -> str:
+    """
+    Given a string and it's "true" length, this function will return a new string
+    that replaces all of the spaces of the input string with '%20'.
+    Precondition(s):
+    - length of s <= true_length
+    :param s: the original string to 'urlify'
+    :param true_length: since s may have additional characters, we focus on true_length instead of actual s length
+    :return: string with each space from s replaced with '%20'
+    """
+    # Below link goes over string concat efficiency in python (I will use method 4
+    # https://waymoot.org/home/python_string/
+    output = []
+    for i, c in enumerate(s):
+        if i == true_length:
```

Comment:
note that this will almost never be true! if `true_length` is equal to `len(s)`, `i` is at most `len(s) - 1`.

the reason why `true_length` is provided is that this assumes that you are coding in C/C++ and you will be modifying the string in-place, but this is not true in Java or Python (hence the weird comment about the character array thing).

if you still want to honor it (e.g. for the case where you have an "ignore this" suffix), do you think you can do it without introducing an additional branch?

Reply:
Oh ok, I see.  I'll add another function without true_length.  Then I will explain in the docstring why true_length is omitted.

Reply:
sure, and if you do end up keeping true_length, try to think of a way to _not_ have an explicit `if i == true_length` check. it's possible and Pythonic.

Reply:
Oh just saw this on refresh, will attempt pythonic way for that now

Reply:
```python
for c in s[0:true_length]
```